homeassistant/components/lock/bmw_connected_drive.py

```diff
@@ -0,0 +1,122 @@
+from homeassistant.components.lock import LockDevice
+from homeassistant.const import STATE_LOCKED, STATE_UNLOCKED
+
+DEPENDENCIES = ['bmw_connected_drive']
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the BMW Connected Drive lock."""
+    accounts = hass.data[BMW_DOMAIN]
+    _LOGGER.debug('Found BMW accounts: %s',
+                  ', '.join([a.name for a in accounts]))
+    devices = []
+    for account in accounts:
+        for vehicle in account.account.vehicles:
+            device = BMWLock(account, vehicle, 'lock', 'BMW lock')
+            devices.append(device)
+    add_devices(devices)
+
+
+class BMWLock(LockDevice):
+    """Representation of a BMW vehicle lock."""
+
+    def __init__(self, account, vehicle, attribute: str, sensor_name):
+        """Initialize the lock."""
+        self._account = account
+        self._vehicle = vehicle
+        self._attribute = attribute
+        self._name = sensor_name
+        self._state = None
+
+    @property
+    def should_poll(self):
+        """Do not poll this class.
+
+        Updates are triggered from BMWConnectedDriveAccount.
+        """
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the lock."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the lock."""
+        vehicle_state = self._vehicle.state
+        return {
+            'last_update': vehicle_state.timestamp,
+            'car': self._vehicle.modelName,
+            'door_lock_state': vehicle_state.door_lock_state.value
+        }
+
+    @property
+    def is_locked(self):
+        """Return true if lock is locked."""
+        return self._state == STATE_LOCKED
+
+    @asyncio.coroutine
+    def async_lock(self, **kwargs):
+        """Lock the car."""
+        from bimmer_connected.remote_services import ExecutionState
+
+        _LOGGER.info("%s: locking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_lock()
+        self._state = STATE_LOCKED
+        self.async_schedule_update_ha_state()
+        while status.state != ExecutionState.EXECUTED:
+            status = self._vehicle.remote_services.get_remote_service_status()
+            yield from asyncio.sleep(1)
+        _LOGGER.info("%s: doors locked", self._vehicle.modelName)
+        yield from asyncio.sleep(10)
+        self.hass.async_add_job(self._account.update)
+
+    @asyncio.coroutine
+    def async_unlock(self, **kwargs):
+        """Unlock the car."""
+        from bimmer_connected.remote_services import ExecutionState
+
+        _LOGGER.info("%s: unlocking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_unlock()
+        self._state = STATE_UNLOCKED
+        self.async_schedule_update_ha_state()
```

Comment:
See above.

Reply:
Locking/unlocking the car doesn't trigger a state update.